### PR TITLE
Feat: scoped variables config

### PR DESCRIPTION
**Pull request by AbhishekA1509**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

This PR includes the entire flow to upload scoped variables.
The flow includes:
- Upload view when Super Admin has saved no variables.
- Editor View when Super Admin wants to edit the variables file
- Review view showing diff of currently saved variables and edited variables.
- List View showing list of variables with name and description with Search Bar.
- Read only view where Super Admin can see the saved file.
- Enable the feature based on flag.

Snapshots:
- Variable List
Screenshot 2023-09-01 at 3 19 22 PM

- Saved Editor view:
Screenshot 2023-09-01 at 3 17 38 PM
- Review file mode:
Screenshot 2023-09-01 at 3 18 14 PM
- Edit file mode:
Screenshot 2023-09-01 at 3 18 41 PM

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/common/formFields/Widgets/Widgets.tsx [hunk 1]
```diff
@@ -328,9 +340,16 @@ export const StyledProgressBar = ({ resetProgress, updateProgressValue }: Styled
                 clearInterval(progressTimer)
             }
         }
-    }, [resetProgress])
+    }, [resetProgress, progress])
 
-    return <progress className="styled-progress-bar" value={progressValue} max={100} />
+    return (
+        <progress
+            className={`styled-progress-bar ${classes ?? ''}`}
```
Comment (shivani170) [inline src/components/common/formFields/Widgets/Widgets.tsx:347]: Please use double quotes for all html elements

File: src/components/scopedVariables/index.tsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
+import React, { useEffect, useState } from 'react'
```
Comment (shivani170) [inline src/components/scopedVariables/index.tsx:1]: Rename this to ScopedVariable.tsx

File: src/components/scopedVariables/mocks.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
+export const validScopedVariablesData = {
+    result: {
```
Comment (shivani170) [inline src/components/scopedVariables/mocks.ts:2]: please include the status code as well in the mock

File: src/components/scopedVariables/styles.scss [hunk 4]
```diff
@@ -0,0 +1,62 @@
+.descriptor-container__upload-button:hover {
```
Comment (shivani170) [inline src/components/scopedVariables/styles.scss:1]: Reduce the css

File: src/components/scopedVariables/ScopedVariablesEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+import { ReactComponent as ICArrowRight } from '../../assets/icons/ic-arrow-right.svg'
+import { ReactComponent as ICPencil } from '../../assets/icons/ic-pencil.svg'
+import {
+    PARSE_ERROR_TOAST_MESSAGE,
+    SAVE_ERROR_TOAST_MESSAGE,
+    SAVE_SUCCESS_TOAST_MESSAGE,
+    GET_SCOPED_VARIABLES_ERROR,
+} from './constants'
+
+const ScopedVariablesEditor = ({
```
Comment (shivani170) [inline src/components/scopedVariables/ScopedVariablesEditor.tsx:25]: make it a functional component

File: src/components/scopedVariables/Descriptor.tsx [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        if (readFile) {
+            readFile(e.target.files![0], validator, ReadFileAs.TEXT)
+        }
+    }
+
+    return (
+        <>
+            <div
+                className="pt-16 pb-16 pl-20 pr-20 flex column dc__align-self-stretch bcn-0 dc__content-space dc__align-start dc__border-bottom"
+                style={children ? { padding: '16px 20px 8px 20px', borderBottom: 'none' } : {}}
```
Comment (shivani170) [inline src/components/scopedVariables/Descriptor.tsx:23]: Can use classname instead

File: src/components/scopedVariables/Descriptor.tsx [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+                            placement="right"
+                            Icon={QuestionFilled}
+                            heading={DEFAULT_TITLE}
+                            infoText={DEFAULT_DESCRIPTION}
+                            showCloseButton={true}
+                            trigger="click"
+                            interactive={true}
+                        >
+                            <button className="p-0 h-20 dc__no-background dc__no-border dc__outline-none-imp">
+                                <ICHelpOutline width={20} height={20} />
```
Comment (shivani170) [inline src/components/scopedVariables/Descriptor.tsx:41]: Can use icon-dim-20

File: src/components/scopedVariables/Descriptor.tsx [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+import { TippyCustomized, TippyTheme } from '@devtron-labs/devtron-fe-common-lib'
+import { validator } from './utils/helpers'
+import { DescriptorI, ReadFileAs } from './types'
+import { ReactComponent as ICHelpOutline } from '../../assets/img/ic-help-outline.svg'
+import { ReactComponent as QuestionFilled } from '../../assets/icons/ic-help.svg'
+import { ReactComponent as ICUpload } from '../../assets/icons/ic-upload-blue.svg'
+import { DEFAULT_DESCRIPTION, DEFAULT_TITLE } from './constants'
+import ScopedVariablesInput from './ScopedVariablesInput'
+
+const Descriptor = ({ children, showUploadButton, readFile }: DescriptorI) => {
```
Comment (shivani170) [inline src/components/scopedVariables/Descriptor.tsx:11]: Replace all the methods with function

File: src/components/scopedVariables/Descriptor.tsx (hunk 7 above)
Comment (AbhishekA1509) [reply src/components/scopedVariables/Descriptor.tsx]: Done

File: src/components/scopedVariables/Descriptor.tsx (hunk 8 above)
Comment (AbhishekA1509) [reply src/components/scopedVariables/Descriptor.tsx]: Done

File: src/components/scopedVariables/ScopedVariablesEditor.tsx (hunk 5 above)
Comment (AbhishekA1509) [reply src/components/scopedVariables/ScopedVariablesEditor.tsx]: Done

File: src/components/scopedVariables/Descriptor.tsx (hunk 6 above)
Comment (AbhishekA1509) [reply src/components/scopedVariables/Descriptor.tsx]: Done

File: src/components/scopedVariables/index.tsx (hunk 2 above)
Comment (AbhishekA1509) [reply src/components/scopedVariables/index.tsx]: Done

File: src/components/scopedVariables/styles.scss (hunk 4 above)
Comment (AbhishekA1509) [reply src/components/scopedVariables/styles.scss:1]: Done

**Review by shivani170 — APPROVED** [review]

File: src/components/common/hooks/types.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
+export enum ReadFileAs {
```
[the remark this answers is not in the document]
Comment (AbhishekA1509) [reply src/components/common/hooks/types.ts:1]: Done

File: src/components/common/hooks/FileReader.ts [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        return () => {
+            reader.removeEventListener('load', handleFileRead)
+            reader.removeEventListener('error', handleFileError)
+            reader.removeEventListener('progress', handleFileProgress)
+        }
+    }, [])
+
+    useEffect(() => {
+        // doing == since we want to check for null and undefined
+        if (!fileData || fileData.data == null) return
```
Comment (vivek-devtron) [inline src/components/common/hooks/FileReader.ts:22]: Combined these conditions

File: src/components/common/hooks/FileReader.ts [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        if (!validator) return
+        const { status, message } = validator(fileData)
+        setStatus(() => ({
+            message,
+            status,
+        }))
+        if (status) {
+            setProgress(() => 100)
```
Comment (vivek-devtron) [inline src/components/common/hooks/FileReader.ts:30]: We can directly assign the value directly. No need to create inline method if we are not using the prev value. Handle it everyWhere

File: src/components/common/hooks/types.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+    TEXT = 'text',
+    DATA_URL = 'dataUrl',
+    BINARY_STRING = 'binaryString',
+    ARRAY_BUFFER = 'arrayBuffer',
+}
+
+export interface FileReaderStatusInterface {
+    status: boolean | 'loading'
```
Comment (vivek-devtron) [inline src/components/common/hooks/types.ts:9]: Please fix this type

File: src/components/scopedVariables/DescriptorSearchBar.tsx [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+            onSearch('')
+        }
+        setSearchText(e.target.value)
+    }
+
+    return (
+        <div
+            className={`dc__border-radius-4-imp flexbox pt-5 pr-10 pb-5 pl-8 dc__gap-8 w-250 dc__bg-n50 dc__align-items-center dc__border ${containerClass}`}
+        >
+            <div
```
Comment (vivek-devtron) [inline src/components/scopedVariables/DescriptorSearchBar.tsx:34]: Please change it to SVG

File: src/components/scopedVariables/types.tsx [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+export interface VariableSpecInterface {
+    description?: string
+    name: string
+    values: ValueInterface[]
+}
+
+export interface ValueInterface {
+    category: 'ApplicationEnv' | 'Application' | 'Env' | 'Cluster' | 'Global'
```
Comment (vivek-devtron) [inline src/components/scopedVariables/types.tsx:79]: we should use VariableCategories here

File: src/components/common/hooks/FileReader.ts (hunk 10 above)
Comment (AbhishekA1509) [reply src/components/common/hooks/FileReader.ts]: Done

File: src/components/common/hooks/FileReader.ts (hunk 11 above)
Comment (AbhishekA1509) [reply src/components/common/hooks/FileReader.ts]: Done

File: src/components/common/hooks/types.ts (hunk 12 above)
Comment (AbhishekA1509) [reply src/components/common/hooks/types.ts]: Done

File: src/components/scopedVariables/ScopedVariablesEditor.tsx [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+            <div className="flexbox-col p-8 dc__align-start dc__gap-16 dc__align-self-stretch dc__window-bg flex-grow-1 dc__no-shrink">
+                <div className="flexbox-col dc__content-space dc__align-start flex-grow-1 dc__no-shrink dc__align-self-stretch dc__border-radius-4-imp dc__border">
+                    <div className="flexbox pt-8 pb-8 pl-12 pr-12 bcn-0 dc__border-bottom dc__gap-16 dc__align-self-stretch dc__align-start dc__top-radius-4">
+                        {setShowEditView ? (
+                            <p className="flex-grow-1 dc__no-shrink cn-9 fs-13 fw-4 lh-20 m-0 dc__ellipsis-right">
+                                {showSaveView ? 'Review' : 'Edit'} <span style={{ fontWeight: 700 }}>Variables</span>
+                            </p>
+                        ) : (
+                            <p className="flex-grow-1 dc__no-shrink cn-9 fs-13 fw-4 lh-20 m-0 dc__ellipsis-right">
+                                Upload{' '}
```
Comment (vivek-devtron) [inline src/components/scopedVariables/ScopedVariablesEditor.tsx:126]: PLease remove {' '}

File: src/components/scopedVariables/ScopedVariablesEditor.tsx [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+                                <span style={{ fontWeight: 700 }}>{name?.split('.').slice(0, -1).join('.')}</span>
```
Comment (vivek-devtron) [inline src/components/scopedVariables/ScopedVariablesEditor.tsx:127]: we can use class 'fw-7'

File: src/components/scopedVariables/ScopedVariablesEditor.tsx [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+                            </p>
+                        )}
+
+                        <Tippy
+                            className="default-tt"
+                            arrow
+                            placement="top"
+                            content={
+                                <div>
```
Comment (vivek-devtron) [inline src/components/scopedVariables/ScopedVariablesEditor.tsx:136]: Why is this div required?

File: src/components/scopedVariables/ScopedVariablesEditor.tsx (hunk 16 above)
Comment (AbhishekA1509) [reply src/components/scopedVariables/ScopedVariablesEditor.tsx]: Done

File: src/components/scopedVariables/ScopedVariablesEditor.tsx (hunk 17 above)
Comment (AbhishekA1509) [reply src/components/scopedVariables/ScopedVariablesEditor.tsx]: Fixed

File: src/components/scopedVariables/ScopedVariablesEditor.tsx (hunk 16 above)
Comment (AbhishekA1509) [reply src/components/scopedVariables/ScopedVariablesEditor.tsx]: Done

File: src/components/scopedVariables/DescriptorSearchBar.tsx (hunk 13 above)
Comment (AbhishekA1509) [reply src/components/scopedVariables/DescriptorSearchBar.tsx]: Done

File: src/components/scopedVariables/types.tsx (hunk 14 above)
Comment (AbhishekA1509) [reply src/components/scopedVariables/types.tsx]: Done

**Review by vivek-devtron — APPROVED** [review]

**Pull request merged**